staticaddr/sql_store.go

```diff
@@ -0,0 +1,146 @@
+package staticaddr
+
+import (
+	"context"
+	"errors"
+
+	"github.com/btcsuite/btcd/btcec/v2"
+	"github.com/jackc/pgx/v4"
+	"github.com/lightninglabs/loop/loopdb"
+	"github.com/lightninglabs/loop/loopdb/sqlc"
+	"github.com/lightningnetwork/lnd/keychain"
+)
+
+// SqlStore is the backing store for static addresses.
+type SqlStore struct {
+	baseDB *loopdb.BaseDB
+}
+
+// NewSqlStore constructs a new SQLStore from a BaseDB. The BaseDB is agnostic
+// to the underlying driver which can be postgres or sqlite.
+func NewSqlStore(db *loopdb.BaseDB) *SqlStore {
+	return &SqlStore{
+		baseDB: db,
+	}
+}
+
+// ExecTx is a wrapper for txBody to abstract the creation and commit of a db
+// transaction. The db transaction is embedded in a `*sqlc.Queries` that txBody
+// needs to use when executing each one of the queries that need to be applied
+// atomically.
+func (s *SqlStore) ExecTx(ctx context.Context, txOptions loopdb.TxOptions,
+	txBody func(queries *sqlc.Queries) error) error {
+
+	// Create the db transaction.
+	tx, err := s.baseDB.BeginTx(ctx, txOptions)
+	if err != nil {
+		return err
+	}
+
+	// Rollback is safe to call even if the tx is already closed, so if the
+	// tx commits successfully, this is a no-op.
+	defer func() {
+		err := tx.Rollback()
+		switch {
+		// If the tx was already closed (it was successfully executed)
+		// we do not need to log that error.
+		case errors.Is(err, pgx.ErrTxClosed):
+			return
+
+		// If this is an unexpected error, log it.
+		case err != nil:
+			log.Errorf("unable to rollback db tx: %v", err)
+		}
+	}()
+
+	if err := txBody(s.baseDB.Queries.WithTx(tx)); err != nil {
+		return err
+	}
+
+	// Commit transaction.
+	return tx.Commit()
+}
+
+// CreateStaticAddress creates a static address record in the database.
+func (s *SqlStore) CreateStaticAddress(ctx context.Context,
+	addrParams *AddressParameters) error {
+
+	createArgs := sqlc.CreateStaticAddressParams{
+		ClientPubkey:    addrParams.ClientPubkey.SerializeCompressed(),
+		ServerPubkey:    addrParams.ServerPubkey.SerializeCompressed(),
+		Expiry:          int32(addrParams.Expiry),
```

Comment:
Make sure `addrParams.Expiry` (uint32) does not overflow `int32`.
Maybe similar checks are needed for other fields (KeyLocator.Family, KeyLocator.Index, ProtocolVersion). Maybe this should be checked in higher layers, where the values are validated.